river/cluster/denstream.py

```diff
@@ -43,411 +44,424 @@ class DenStream(base.Clusterer):
-                for i in range(self.dim)
-            }
 
-    @property
-    def weighted_linear_sum(self):
-        weighted_linear_sum = {
-            i: (2 ** (-self.current_time * self.decaying_factor)) * self.IWLS[i]
-            for i in range(self.dim)
-        }
-        return weighted_linear_sum
-
-    @property
-    def weighted_squared_sum(self):
-        weighted_squared_sum = {
-            i: (2 ** (-self.current_time * self.decaying_factor)) * self.IWSS[i]
-            for i in range(self.dim)
-        }
-        return weighted_squared_sum
-
-    @property
-    def weight(self):
-        weight = (
-            2 ** (-self.current_time * self.decaying_factor)
-        ) * self.initial_weight
-        return weight
-
-    @property
-    def center(self):
-        center = {i: self.weighted_linear_sum[i] / self.weight for i in range(self.dim)}
+        self.N = 1
+        self.feature_keys = set(x.keys())
+        self.liner_sum = x
+        self.squared_sum = {i: (x_val * x_val) for i, x_val in x.items()}
+
+    def calc_norm_cf1_cf2(self, fading_factor):
+        # |CF1| and |CF2| in the paper
+        sum_of_squares_cf1 = 0
+        sum_of_squares_cf2 = 0
+        for key in self.liner_sum.keys():
+            val_ls = self.liner_sum[key]
+            val_ss = self.squared_sum[key]
+            sum_of_squares_cf1 += fading_factor * val_ls * fading_factor * val_ls
+            sum_of_squares_cf2 += fading_factor * val_ss * fading_factor * val_ss
+        # return |CF1| and |CF2|
+        return math.sqrt(sum_of_squares_cf1), math.sqrt(sum_of_squares_cf2)
+
+    def calc_weight(self, timestamp):
+        return self._weight(self.fading_function(timestamp - self.last_edit_time))
+
+    def _weight(self, fading_factor):
+        return self.N * fading_factor
+
+    def calc_center(self, timestamp):
+        ff = self.fading_function(timestamp - self.last_edit_time)
+        weight = self._weight(ff)
+        center = {key: (ff * val) / weight for key, val in self.liner_sum.items()}
         return center
 
-    @property
-    def radius(self):
-        radius = math.sqrt(
-            abs(
-                self.norm(self.weighted_squared_sum) / self.weight
-                - (self.norm(self.weighted_linear_sum) / self.weight) ** 2
-            )
-        )
+    def calc_radius(self, timestamp):
+        ff = self.fading_function(timestamp - self.last_edit_time)
+        weight = self._weight(ff)
+        norm_cf1, norm_cf2 = self.calc_norm_cf1_cf2(ff)
+        diff = (norm_cf2 / weight) - ((norm_cf1 / weight) ** 2)
+        radius = math.sqrt(diff) if diff > 0 else 0
         return radius
 
-    def add(self, cluster):
-        assert self.dim == cluster.dim
+    def insert(self, x, timestamp):
+        if set(x.keys()) != self.feature_keys:
+            raise KeyError(
+                f"Features mismatch. Expected {self.feature_keys}",
+                f"got {set(x.keys())}",
+            )
```

Comment:
Can't the union of both key sets be used?

Reply:
Interesting, although I am not completely sure how a new feature would fare alongside the existing metrics? I am not saying no, I am just thinking out loud.

Reply:
I don't know either haha

But still, I think it's better to keep processing instead of raising an error. What do you think, @MaxHalford?

Reply:
@jacobmontiel @smastelini IMO, we can have two solutions for this: One is either to raise a 
```python 
KeyError
```
or to simply ignore it and continue, with a 
```python
pass
```
Personally, I would prefer the second way, in the sense that it would not break our stream for one small error.

Reply:
Added support for emerging features in c7ec6f9a8ae5cb5ffdbf5d9d6630c6105d8cfc60